### PR TITLE
Add recipe for useSubscription hook

### PR DESCRIPTION
**Pull request by bvaughn**

### CODE REVIEW

File: content/docs/hooks-recipes.md [hunk 1]
```diff
@@ -10,3 +10,157 @@ prev: hooks-faq.html
 This is a collection of recipes that may help you start "thinking in Hooks". Feel free to play with them or copy and paste them into your project or tweak them to your needs.
 
 ## Data Fetching
+
+## Subscriptions
+
+Sometimes React components need to subscribe to an external data source. The `useSubscription` hook below shows how to do this in a way that is compatibile with React's upcoming concurrent rendering mode.
```
Comment (gaearon) [inline content/docs/hooks-recipes.md]: I would maybe rephrase it so that it doesn't feel like CM is the *problem* here.

Maybe:

> Sometimes React components need to subscribe to an external data source. Handling all the possible edges in a performant way can be challenging, which is why people often use dedicated third-party libraries for connecting to data. The `useSubscription` Hook below simplifies that. It lets you subscribe to an external data source while handling all the edge cases. It is also compatible with the upcoming concurrent rendering mode.

File: content/docs/hooks-recipes.md [hunk 2]
```diff
@@ -10,3 +10,157 @@ prev: hooks-faq.html
...
+
+> Note
+>
+> Subscriptions may not be the best solution for your use case. For example:
+> * Redux/Flux stores should use the [context API](/docs/context.html) instead.
```
Comment (gaearon) [inline content/docs/hooks-recipes.md]: We've gotten flak for this recommendation because it led to React Redux v6 rewrite which has to back out of context now (not fast enough for some extreme use cases). I'd omit or rephrase this.

File: content/docs/hooks-recipes.md [hunk 3]
```diff
@@ -10,3 +10,157 @@ prev: hooks-faq.html
...
+> * Complex libraries like Relay/Apollo should manage subscriptions manually with the same technique demonstrated below.
+> * In the future, I/O subscriptions (e.g. notifications) that update infrequently should use [`react-cache`](https://github.com/facebook/react/blob/master/packages/react-cache/README.md).
```
Comment (gaearon) [inline content/docs/hooks-recipes.md]: We should have no mention of react-cache in official doc. We might not even keep the name. It's also confusing to read about the future in a recipes section — people want to do something today.

File: content/docs/hooks-recipes.md (hunk 2 above)
Comment (bvaughn) [reply content/docs/hooks-recipes.md]: Huh. Fair enough, although that's still in our `create-subscription` guidelines: https://github.com/facebook/react/tree/master/packages/create-subscription#when-should-you-not-use-this

File: content/docs/hooks-recipes.md [hunk 4]
```diff
@@ -10,3 +10,157 @@ prev: hooks-faq.html
...
+
+Although `useSubscription` guarantees correctness, it achieves this by [sometimes de-opting to synchronous mode](https://github.com/facebook/react/issues/13186#issuecomment-403959161). This is an inherent limitation of storing state outside of React's managed state queue and rendering in response to a change event. For **full compatibility** with concurrent rendering, including both **time-slicing** and **React Suspense**, the suggested longer-term solution is to move to one of the patterns described in the note above.
```
Comment (gaearon) [inline content/docs/hooks-recipes.md]: People don't know what "de-opting" or "synchronous mode" is here. We didn't introduce these concepts anywhere in the docs except the blog (which isn't a required reading for docs). I'd omit details about any future features here, and just say it is compatible with Strict Mode (which *is* in the docs).

**Review by gaearon — CHANGES_REQUESTED** [review]

File: content/docs/hooks-recipes.md (hunk 2 above)
Comment (gaearon) [reply content/docs/hooks-recipes.md]: Yes, and we should have fixed it there too. There's a task but nobody got to it yet. :-)

**Comment by bvaughn** [comment]
Back to you.

**Review by gaearon — APPROVED** [review]
